src/GamePlayer.cpp

```diff
@@ -826,45 +826,50 @@ void GamePlayer::FindWarehouseForAllJobs(const Job job)
 
 Ware* GamePlayer::OrderWare(const GoodType ware, noBaseBuilding* goal)
 {
-    /// Gibt es ein Lagerhaus mit dieser Ware?
-    nobBaseWarehouse* wh = FindWarehouse(*goal, FW::HasMinWares(ware, 1), false, true);
-
-    if(wh)
+    if(emergency)
     {
-        // Prüfe ob Notfallprogramm aktiv
-        if(!emergency)
-            return wh->OrderWare(ware, goal);
-        else
-        {
-            // Wenn Notfallprogramm aktiv nur an Holzfäller und Sägewerke Bretter/Steine liefern
-            if((ware != GD_BOARDS && ware != GD_STONES) || goal->GetBuildingType() == BLD_WOODCUTTER
-               || goal->GetBuildingType() == BLD_SAWMILL)
-                return wh->OrderWare(ware, goal);
-            else
-                return NULL;
-        }
-    } else // no warehouse can deliver the ware -> check all our wares for lost wares that might match the order
+        if(ware == GD_STONES && goal->GetBuildingType() != BLD_SAWMILL)
+            return NULL;
+
+        if(ware == GD_BOARDS && (goal->GetBuildingType() != BLD_WOODCUTTER && goal->GetBuildingType() != BLD_SAWMILL))
+            return NULL;
+    }
+
+    // Look for available ware in a warehouse
+    unsigned int nearestWarehouse = std::numeric_limits<unsigned>::max();
+    nobBaseWarehouse* wh = FindWarehouse(*goal, FW::HasMinWares(ware, 1), false, true, &nearestWarehouse);
+
+    // Look for ware waiting on the road or on the way to storehouse (if addon is enabled)
+    unsigned nearestWare = std::numeric_limits<unsigned>::max();
+    Ware* bestWare = NULL;
+    const bool improvedWareOrder = gwg->GetGGS().isEnabled(AddonId::IMPROVED_WARE_ORDER);
+    if(!wh || improvedWareOrder)
     {
-        unsigned bestLength = std::numeric_limits<unsigned>::max();
-        Ware* bestWare = NULL;
         BOOST_FOREACH(Ware* curWare, ware_list)
         {
-            if(curWare->IsLostWare() && curWare->type == ware)
+            noBaseBuilding* wareGoal = curWare->GetGoal();
+            // Ware is redirectable if its lost (has no goal) or its goal is a storehouse
+            if(curWare->type == ware && (curWare->IsLostWare() || (improvedWareOrder && wareGoal->GetBuildingType() == BLD_STOREHOUSE)))
             {
-                // got a lost ware with a road to goal -> find best
                 unsigned curLength = curWare->CheckNewGoalForLostWare(goal);
```

Comment:
Maybe add an extra param to pass the maximum distance as we (may) already have a distance (for other ware or warehouse)

Note that this also ONLY works for wares currently at flags. This is better than nothing but not optimal. Maybe rework this in a follow up.

Reply:
Good idea to add a max distance and give its current distance from warehouse. FindPathForWareOnRoads already supports max distance, so its just adding NULL and the new parameter.

About the Ware not at a flag:
Shouldnt it be enough, to use the following code:

`
    noRoadNode* startNode = location;

    if (!IsWaitingAtFlag()) {
        startNode = location->GetNeighbour(Direction::fromInt(next_dir));
    }
` 
And change the other location variables to startNode? I still need to check if i need to call a Carrier, but before this i want to get feedback on my idea.

Reply:
If the ware is being carried, then yes this should be enough. IIRC you can even just set the new goal when the ware is carried and it will calculate the path, once it reaches the next flag.